ES-34 Add logic for password reset functionality

tk1ng opened the pull request:
## Description
Adds logic to handle password reset.

### There are three steps that happen on the backend for the process:

1. Client requests password reset, which is handled by `/auth/forgot-password`  endpoint

    - This endpoint prompts the reset email to be sent to user. Email contains link that includes parameters with a token
    for verification and the URL for the reset page that user is redirected to for updating password.

2.  User receives email and clicks reset link that routes to `/auth/confirm`.
    - This endpoint verifies that the token from the email is valid. If valid, sets a cookie beginning with `sb-` in the browser for authenticating the user and then redirects to the reset page
    - If is is not valid, the endpoint redirects to an error page

3. Finally, the reset form submit should call `auth/update-password`
    - This endpoint, verifies cookie is present and valid and updates the user password. Then redirects to login page

Explored using `createServerClient` to create an instance of supabase that handles setting cookies for authenticating user for the password reset action. This logic was added to `authClient/index.ts` and the constructor is called from the `verifyOtp` and `updatePassword` functions defined in `authFunctions.ts`

## How to Test
*Before beginning test, run `npm i` from the feature branch. This will install the `@supabase/ssr` package that is used in this logic.

To test this, a page is needed to stand in for the reset password page. 
1. Add the follow code to the top of the `auth.ts` file just below where `router` is initialized:
    `router.use(express.urlencoded({ extended: true })); `
    
    The form that is rendered to test this logic is sending the password data in URL-encoded form, so this is used to parse the URL-encoded data to be able to access the data from the request body to simulate how the data is passed from the frontend.
2. Add the following test route to the `auth.ts` file that renders a simple reset form
```
    router.get('/reset-test', (req, res) => {
    res.send(`
        <!DOCTYPE html>
        <html lang="en">
        <head>
            <meta charset="UTF-8">
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <title>Reset Password</title>
            <style>
                body {
                    font-family: Arial, sans-serif;
                    background-color: #f4f4f4;
                    display: flex;
                    justify-content: center;
                    align-items: center;
                    height: 100vh;
                    margin: 0;
                }
                .reset-form {
                    background-color: #fff;
                    padding: 20px;
                    border-radius: 8px;
                    box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
                    width: 300px;
                }
                .reset-form h2 {
                    margin-bottom: 20px;
                    font-size: 24px;
                    color: #333;
                }
                .reset-form input[type="password"] {
                    width: 93%;
                    padding: 10px;
                    margin-bottom: 20px;
                    border: 1px solid #ccc;
                    border-radius: 4px;
                    font-size: 16px;
                }
                .reset-form button {
                    width: 100%;
                    padding: 10px;
                    background-color: #6638f0;
                    color: #fff;
                    border: none;
                    border-radius: 4px;
                    font-size: 16px;
                    cursor: pointer;
                }
                .reset-form button:hover {
                    background-color: #6638f0;
                }
            </style>
        </head>
        <body>
            <div class="reset-form">
                <h2>Reset Password</h2>
                <form action="http://localhost:3000/auth/update-password" method="POST">
                    <label for="newPassword">New Password:</label>
                    <input type="password" id="newPassword" name="newPassword" required>
                    <button type="submit">Reset Password</button>
                </form>
            </div>
        </body>
        </html>
    `);
});
```
3. Start server and use Thunder Client or HTTP client tool of choice to call `/auth/forgot-password`, sending a test email from db as `email` param in body. (Aim to use a tenant that has a value set for the `user_id` field on the tenants table)
4. Access reset email and click link
    - to verifiy cookie is set, check dev tools > application > cookies for cookie beginning with `sb`
    ![CleanShot 2025-03-07 at 09 54 17@2x](https://github.com/user-attachments/assets/59c0e918-a2e0-4c45-8d6e-8121a39232a1)
5. Add new password and submit form
6. Success message returned (this will be updated to redirect to the corresponding route on the frontend for login page)
![CleanShot 2025-03-09 at 20 57 46@2x](https://github.com/user-attachments/assets/08ac76b9-7362-4475-8eaa-f6d03839bb14)

**Testing Error**
To test for error, delete the cookie from the browser before submitting the form. This should return an error that user is not authorized and unable to reset password.
![CleanShot 2025-03-07 at 09 57 31@2x](https://github.com/user-attachments/assets/85d6f4b5-8ee9-45d8-93d6-2f670f742f7a)

**Visual of reset flow**
![pass_reset_flow](https://github.com/user-attachments/assets/8b9024ce-63a3-4e17-8d46-46d2e4c82af3)

tk1ng committed:
feat(auth): add logic for password reset functionality

tk1ng committed:
Merge branch 'main' into ES-34-password-reset-functionality

CloudsProgram commented:
I tried running the server and it is throwing the following errors:
![image](https://github.com/user-attachments/assets/ff979542-188f-43ea-b828-16d11361c95e)

tk1ng commented:
Oh! I forgot to add in the instructions that a new package was added that will need to be installed through npm install

CloudsProgram commented:
I was able to receive the confirmation email, but after clicking on the link in the email it produces an error page.
and the server crashed and produced the following in the console:
![image](https://github.com/user-attachments/assets/fd175e6a-ef74-467a-8e81-69acd997c320)

tk1ng commented:
@CloudsProgram do you have a screenshot of the full call stack trace?

CloudsProgram commented:
```js
TypeError: context.res.appendHeader is not a function
    at C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\src\authClient\index.ts:21:33
    at Array.forEach (<anonymous>)
    at setAll (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\src\authClient\index.ts:20:30)
    at applyServerStorage (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\@supabase\ssr\src\cookies.ts:469:9)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Object.callback (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\@supabase\ssr\src\createServerClient.ts:196:7)
    at async C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEn                             ^
TypeError: context.res.appendHeader is not a function
    at C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\src\authClient\index.ts:21:33
    at Array.forEach (<anonymous>)
    at setAll (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\src\authClient\index.ts:20:30)
    at applyServerStorage (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\@supabase\ssr\src\cookies.ts:469:9)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Object.callback (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\@supabase\ssr\src\createServerClient.ts:196:7)
    at async C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\@supabase\auth-js\src\GoTrueClient.ts:2040:11
    at async Promise.all (index 1)
    at async SupabaseAuthClient._notifyAllSubscribers (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\@supabase\auth-js\src\GoTrueClient.ts:2046:7)
    at async SupabaseAuthClient.verifyOtp (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\@supabase\auth-js\src\GoTrueClient.ts:783:9)
```

tk1ng committed:
fix: set cookies using setHeader to align with express

tk1ng commented:
The reason for that error is that I was using `appendHeader` but Express uses `setHeader`.  I made another commit to update this in the code.

I also updated the instructions for testing to include a note for adding a middleware for testing that parses URL-encoded data to allow for testing with the temporary form shared above. It allows us to mimic how the data will be sent from the frontend.

CloudsProgram commented:
I've sent a POST request (with cookie) to `http://localhost:3000/auth/forgot-password` without a body, and the program crashed with the following
```js
AuthApiError: Password recovery requires an email
    at handleError (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\@supabase\auth-js\src\lib\fetch.ts:102:9)
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async _handleRequest (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\@supabase\auth-js\src\lib\fetch.ts:195:5)      
    at async _request (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\@supabase\auth-js\src\lib\fetch.ts:157:16)
    at async SupabaseAuthClient.resetPasswordForEmail (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\@supabase\auth-js\src\GoTrueClient.ts:1713:14) {
  __isAuthError: true,
  status: 400,
  code: 'validation_failed'
}
::1 - - [10/Mar/2025:04:18:54 +0000] "POST /auth/forgot-password HTTP/1.1" 500 34
Error: Cannot set headers after they are sent to the client
    at ServerResponse.setHeader (node:_http_outgoing:699:11)
    at ServerResponse.header (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\express\lib\response.js:794:10)
    at ServerResponse.send (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\express\lib\response.js:174:12)
    at ServerResponse.json (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\node_modules\express\lib\response.js:278:15)
    at C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\src\routes\auth.ts:55:25
    at Generator.next (<anonymous>)
    at fulfilled (C:\Users\cloud\OneDrive\Desktop\Commit Your Code folder\project\playground\EliteSpace-BackEnd\src\routes\auth.ts:5:58)
    at processTicksAndRejections (node:internal/process/task_queues:105:5) {
  code: 'ERR_HTTP_HEADERS_SENT'
}
[nodemon] app crashed - waiting for file changes before starting...
```

tk1ng commented:
@CloudsProgram Did you pull down the latest update to this branch in your local env and run the most recent test starting with requesting a reset email? Also, you shouldn't have to do anything on your end in terms of adding a cookie.

CloudsProgram commented:
I did pull the most recent update from origin to my local branch. And following the steps that you listed out.

tk1ng commented:
Hmm, I see. It's working on my end. I'm happy to sync tomorrow if you are free to see what could be wrong.

tk1ng committed:
fix: adding return statements in route handler

This is to ensure that the remaining parts of the function do not run if response has been sent

tk1ng committed:
chore: correct spelling of return

eddiemaceda11 reviewed (approved):
Tested with both success and error requests. The entire password reset process is working as intended. Great job on the PR @tk1ng, very detailed and easy to follow.

CloudsProgram commented:
It works now on my end! Looks good to me

Patrick-250 reviewed (approved):
Just tested everything and it works as expected. I was able to receive the reset link and also reset password. Good Job Taryn, well explained too!
